Replacing logxi with simple LogFunc approach

bkono opened the pull request:
This commit removes the dependence on a single logging package. A new
dat/log package is added that allows setting debug, sql, and err
LogFuncs, while also providing an api compatible to logxi for ease of
transition. More work should be done here to nix things like log.Error
the inconsistent string returns, and log.Fatal.

bkono committed:
Replacing logxi with simple LogFunc approach

This commit removes the dependence on a single logging package. A new
dat/log package is added that allows setting debug, sql, and err
LogFuncs, while also providing an api compatible to logxi for ease of
transition. More work should be done here to nix things like log.Error
the inconsistent string returns, and log.Fatal.

bkono commented:
This is one, modestly clean, approach to addressing https://github.com/helloeave/dat/issues/8 mostly in-place, without too many visible changes.

@pascallouisperez Plucked out, and slightly cleaned up, just the logxi / LogFunc related changes. Feedback welcome. I did a quick sample app that used both zap (with the exact code added to the README sample) and log.Printf without issue.

bkono committed:
removing logxi from glide deps

bkono commented:
@pascallouisperez I've gone ahead and started using this branch in a service I'm working on. So far, so good. lmk if you have any suggestions for changes to get it integrated.

pascallouisperez reviewed (commented):
It seems like the pattern used so far for the various logging methods (`Debug`, `Err`, etc.) is more complex than simply deferring to `Printf` allows. See https://github.com/mgutz/logxi/blob/master/v1/textFormatter.go#L76

There seem to be four distinct cases:
- no args: print only message
- one arg: print just the arg after the message
- even number of args: print key:value pairs
- odd number of args: print all args with a warning color

My suggestion to deal with this, would be to introduce a conversion from the current calls (e.g. `log.Err`) to what we intend to be `Sprintf`-compatible.

This conversion can live in `log/logger.go`, and basically would do a straight pass-through if `%` is found in the `msg` string.

We simply need to be careful to invoke this conversion only if we actually will log (i.e. underlying logger is not `nil`.)

How does that sound?

pascallouisperez commented:
@bkono Thanks for the ping, got busy, and this dropped from my list :)

bkono commented:
@pascallouisperez No worries on the ping. :) 

Good feedback, and also indicates I should probably make a few API-ish changes. 

Few things on the way I was looking at the change... 

1 - LogFunc is meant to meet the structured logger interface of logxi/log15/zap.SugaredLogger etc, rather than a Sprintf interface. As defined, logxi.Debug, sugar.Debugw, etc all are directly usable without adapters. Using stdlog.Printf as a default was actually a shortcut / miss on my part, where a func wrapping stdlog.Println would have been more accurate and aligned with the structured approach.
2 - dat/log is purely for dat's internal usage. Might actually make sense to move the actual impl down into an internal/ package to avoid it being depended on, and migrate the interface & Set* to dat/dat. 
3 - Given 2, you can see in the change set I actually replaced the very few instances (mostly accidental usage of the stdlog instead of logxi packages, and Fatals) of Printf logging to be fmt.Sprintf'd in-line.

Since I had been thinking of this as purely for dat's usage to become a good go package citizen and not force a logger on anyone, I think we should try hard to steer away from introducing any real complexity, or perf bottlenecks, into the logger. I'd expect everyone consuming it to either run with using the stdlog, which I should provide a cleaner way to use, or to provide an adapter func for whatever logger they are already using (logrus & zap are looking to be the most common from my recent discussions with folks). 

On that note, what do you think about the following tweaks:

1 - Shift the packaging around to make it clear dat/log is just for dat to conveniently provide some logging internally.
2 - Provide something like a dat.StdLogger which uses log.Println instead of Printf and can be set conveniently by a caller with `dat.SetSQLLogger(dat.StdLogger)` if they don't want to bring their own, but do want to turn on SQL or Debug logging. That would make it equally as clean as, for instance, `dat.SetDebugLogger(sugar.Debugw)` for using zap without worrying about a forced sync.
3 - Re-run through the codebase and move any fmt.Sprintf's carried forward to be actual structured logging.

Thoughts?

bkono committed:
Converting log to internal and cleaning up the api

To clarify expected usage, the log package has been moved to internal,
with only the setter and stdlog elements accessible for external
callers. README changes were started as well.

bkono commented:
@pascallouisperez Latest commit is a stab at the above changes. I'll be pulling the updates into a service today to see how they play out.

pascallouisperez commented on internal/log/logger.go:
```diff
@@ -0,0 +1,62 @@
+	ErrorLogFn LogFunc = nil
+)
+
+type LogFunc = func(string, ...interface{})
```
Why is the argument signature `(string, ...interface{})` if the intent is to be compatible with `log.Println` (https://golang.org/pkg/log/#Println)?

bkono replied:
Its actually not for log.Println, because log.Println is not a structured logger. Over the last few years, the go community has mostly settled on the notion that structured loggers (not printf ones) should adhere to either `func(...interface{})` a la go-kit, or `func(string, ...interface{})` a la zap, log15, logxi. I went with the latter because it allows performance benefits like msg level de-duping, and makes the old logxi calls a direct drop in, no adapter func needed, along with zap.Sugared which is the most common structured logging library I've seen in the wild.

pascallouisperez commented on internal/log/logger.go:
```diff
@@ -0,0 +1,62 @@
+package log
```
Let's rename this to `internal_log` (or similar):
- It can then be in `internal_log/` directory (same level as `dat`, `kvs`, etc.)
- It conveys clearly that this is not intended to be external API, and that we're only exporting symbols for re-use internally in other packages
- Avoids clashing with `log` package of stdlib, which should be avoided (https://blog.golang.org/package-names)

bkono replied:
Sure, this is definitely a style thing. I come from this school https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1 where you package dependencies to specifically cause a collision and encourage not directly accessing them. For reference, using log and renaming / shadowing stdlog is actually the common practice among the most popular log interfaces. Examples: https://github.com/go-log/log/blob/master/log/log.go and https://github.com/go-kit/kit/tree/master/log which both cover stdlib log to discourage outside use of it.

Also, we don't want this at the same level as dat/kvs/etc intentionally. The internal/ package is being used here to to create a compile time error if any consumer of dat attempts to import this log package. Reference from official integration in 1.4 https://golang.org/doc/go1.4#internalpackages & [the proposal](https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit)

pascallouisperez replied:
That's great, thanks for all the explanations! Let's leave as is then.

pascallouisperez commented on internal/log/logger.go:
```diff
@@ -0,0 +1,62 @@
+package log
```
Let's rename this to `internal_log` (or similar):
- It can then be in `internal_log/` directory (same level as `dat`, `kvs`, etc.)
- It conveys clearly that this is not intended to be external API, and that we're only exporting symbols for re-use internally in other packages
- Avoids clashing with `log` package of stdlib, which should be avoided (https://blog.golang.org/package-names)

Then, we can avoid renaming `log` to `stdlog` when importing which is more standard.

pascallouisperez reviewed (commented):
This looks super cool, thanks for this!

Let me know how the run in your service works, and then we can merge this.

jaybeecave commented:
bump?
